src/Common/FileSegment.cpp

```diff
@@ -649,15 +607,37 @@ void FileSegment::assertCorrectnessImpl(std::lock_guard<std::mutex> & /* segment
     assert(download_state != FileSegment::State::DOWNLOADED || std::filesystem::file_size(cache->getPathInLocalCache(key(), offset())) > 0);
 }
 
-void FileSegment::assertNotDetached() const
+void FileSegment::throwDetached() const
+{
+    std::lock_guard segment_lock(mutex);
+    throwDetachedUnlocked(segment_lock);
+}
+
+void FileSegment::throwDetachedUnlocked(std::lock_guard<std::mutex> & segment_lock) const
 {
-    if (detached)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "Operation not allowed, file segment is detached");
+    throw Exception(
+        is_forcefully_detached ? ErrorCodes::CACHE_FILE_SEGMENT_IS_DETACHED : ErrorCodes::LOGICAL_ERROR,
+        "Cache file segment is in detached state, operation not allowed. "
+        "It can happen when cache was concurrently dropped with SYSTEM DROP FILESYSTEM CACHE FORCE. "
+        "Please, retry. File segment info: {}", getInfoForLogImpl(segment_lock));
 }
 
-void FileSegment::assertDetachedStatus(std::lock_guard<std::mutex>  & /* segment_lock */) const
+
+void FileSegment::assertNotDetached(std::lock_guard<std::mutex> & segment_lock) const
 {
-    assert(download_state == State::EMPTY || hasFinalizedState());
+    if (is_detached)
+        throwDetachedUnlocked(segment_lock);
+}
+
+void FileSegment::assertDetachedStatus(std::lock_guard<std::mutex> & segment_lock) const
```

Comment:
I don't understand how this name related to function body. What is different from previous one?

Reply:
Added a comment